src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/RefInitializationHoister.cs

```diff
@@ -0,0 +1,240 @@
+                }
+
+            case BoundKind.FieldAccess:
+                {
+                    var field = (BoundFieldAccess)expr;
+                    if (field.FieldSymbol.IsStatic)
+                    {
+                        // the address of a static field, and the value of a readonly static field, is stable
+                        if (refKind != RefKind.None || field.FieldSymbol.IsReadOnly) return expr;
+                        goto default;
+                    }
+                    Debug.Assert(field.ReceiverOpt != null);
+
+                    if (refKind == RefKind.None)
+                    {
+                        goto default;
+                    }
+
+                    var isFieldOfStruct = !field.FieldSymbol.ContainingType.IsReferenceType;
+
+                    var receiver = HoistExpression(
+                        field.ReceiverOpt,
+                        assignedLocal,
+                        refKind: isFieldOfStruct ? refKind : RefKind.None,
+                        sideEffects,
+                        hoistedSymbols,
+                        ref needsSacrificialEvaluation,
+                        hoister,
+                        createHoistedAccess,
+                        arg,
+                        isRuntimeAsync,
+                        isFieldAccessOfStruct: isFieldOfStruct);
+                    if (receiver.Kind != BoundKind.ThisReference && !isFieldOfStruct)
+                    {
+                        needsSacrificialEvaluation = true; // need the null check in field receiver
+                    }
+
+                    return _factory.Field(receiver, field.FieldSymbol);
+                }
+
+            case BoundKind.ThisReference:
+            case BoundKind.BaseReference:
+            case BoundKind.DefaultExpression:
+                return expr;
+
+            case BoundKind.Call:
+                var call = (BoundCall)expr;
+                // NOTE: There are two kinds of 'In' arguments that we may see at this point:
+                //       - `RefKindExtensions.StrictIn`     (originally specified with 'in' modifier)
+                //       - `RefKind.In`                     (specified with no modifiers and matched an 'in' or 'ref readonly' parameter)
+                //
+                //       It is allowed to spill ordinary `In` arguments by value if reference-preserving spilling is not possible.
+                //       The "strict" ones do not permit implicit copying, so the same situation should result in an error.
+                if (refKind != RefKind.None && refKind != RefKind.In)
+                {
+                    Debug.Assert(refKind is RefKindExtensions.StrictIn or RefKind.Ref or RefKind.Out);
+                    Debug.Assert(call.Method.RefKind != RefKind.None);
+                    _factory.Diagnostics.Add(ErrorCode.ERR_RefReturningCallAndAwait, _factory.Syntax.Location, call.Method);
```

Comment:
>ERR_RefReturningCallAndAwait [](http://example.com/codeflow?start=55&length=28)

Do we have a test for this error and similar one below (`ERR_RefConditionalAndAwait`)?

Reply:
Will runtime async verification to `SpillRefBeforeAwaitReordered.

Reply:
Thanks. Given that this case turned out to need a bit of a code change, I would lean towards caution and ensure that all the nooks of the ref hoisting are covered explicitly by runtime-async scenarios.
There are some cases here that I think would be good to make sure we covered:
- the logic regarding static fields above (line 133)
- the logic regarding this/base/default above (line 168)
- the logic regarding constant values below (line 212)

Reply:
constant values is covered 17 different tests in CodeGenAsyncSpillTests.cs alone, not going to list them all.

I added coverage for static fields, which comes through other test classes. It did require new handling.